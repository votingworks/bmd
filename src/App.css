/* Allow selectors to be used multiple times in this file for grouping of concerns. */
/* stylelint-disable no-duplicate-selectors */

/* Reset */
html {
  box-sizing: border-box;
  line-height: 1;
  user-select: none;
}
*,
*::before,
*::after {
  box-sizing: inherit;
}

body {
  margin: 0;
}

table {
  border-collapse: collapse;
}

fieldset {
  margin: 0;
  border: none;
  padding: 0;
}

legend {
  display: block;
}

:link,
:visited {
  color: rgb(0, 0, 238);
}

:focus {
  outline: rgb(77, 144, 254) dashed 0.25rem;
}

/* Foundational */
html,
body,
#root {
  height: 100%;
  overflow: hidden;
}
@media print {
  html,
  body,
  #root {
    height: auto;
    overflow: visible;
  }
}

/* Typography */
html {
  background: #edeff0;
  color: #263238;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
    'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
    sans-serif;
  font-size: 24px; /* Magic number, must synced with JavaScript global defaults */
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}
@media print {
  html {
    background: #ffffff;

    /* Adjust printed ballot font-size */
    font-size: 16px !important; /* stylelint-disable-line declaration-no-important */
  }
}

/* Accessibility Utilities */
.visually-hidden {
  position: absolute !important; /* stylelint-disable-line declaration-no-important */
  width: 1px;
  height: 1px;
  overflow: hidden;
  clip: rect(1px, 1px, 1px, 1px);
}
.visually-hidden a:focus,
.visually-hidden input:focus,
.visually-hidden button:focus {
  position: static;
  width: auto;
  height: auto;
}

/* Utilities */
.print-only {
  display: none;
}
@media print {
  .print-only {
    display: block;
  }
  .no-print {
    display: none;
  }
}

@media print {
  /* Chrome on Linux prints scrollbars */
  ::-webkit-scrollbar {
    display: none;
  }
}

/* stylelint-disable-next-line selector-class-pattern */
html body .cvox_indicator_container {
  /* Extra specificity/!important ensure this rule is followed. */
  display: none !important; /* stylelint-disable-line declaration-no-important */
}

/* Adjust printed ballot paper settings */
@page {
  margin: 0.375in;
  size: letter portrait;
}
