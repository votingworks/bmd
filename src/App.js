import React, { Component } from 'react';
import PrintProvider, { Print, NoPrint } from 'react-easy-print';
import './App.css';


import QrReader from 'react-qr-scanner';

import Ballot from './Ballot';
import PrintableBallot from './PrintableBallot';

var ELECTION = {
  id: "DEMO",
  title: "Demo Election",
  contests: [
    {
      id: "president",
      type: "plurality",
      name: "President",
      options: [
        "Minnie Mouse",
        "Mickey Mouse",
        "Donald Duck"
        ]
    },
    {
      id: "senator",
      type: "plurality",
      name: "Senator",
      options: [
        "Chad Hanging",
        "Lev Ermachine",
        "John Smith"
        ]
    }
  ]
};

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {};
    this.election = ELECTION;
  }

  ballotReady(ballot) {
    this.setState({ballot: ballot});
  }

  print() {
    window.print();
  }

  handleScan(data) {
    if (data!=null) {
      console.log(data);
    }
  }

  handleError() {
  }

  render() {
    let mainContent, printableBallot;

    if (!this.election) {
      printableBallot = (<div></div>);

      const previewStyle = {
        height: 440,
        width: 620
      };

      mainContent = (
        <div align="center">
          <QrReader
            style={previewStyle}
            onError={this.handleError}
            onScan={this.handleScan}
          />
        </div>
      );
    } else {

      if (this.state.ballot) {
        printableBallot = (
          <PrintableBallot election={this.election} ballot={this.state.ballot} />
        );

        mainContent = (
          <div>
            <div style={{paddingLeft: "400px", paddingRight: "400px", paddingBottom: "50px"}}>{printableBallot}</div>
            <button onClick={this.print.bind(this)} style={{fontSize: "2.5em"}}>Print</button>
          </div>

        );
      } else {
        printableBallot = (<div></div>);
        mainContent = (
          <Ballot election={this.election} onBallotReady={this.ballotReady.bind(this)} />
        );
      }
    }


    return (
      <PrintProvider>
        <Print single printOnly name="ballot">
          {printableBallot}
        </Print>

        <NoPrint force>
          <div className="App">
            <img src="./vw-checkmark.png" className="App-logo" alt="logo" />
            {mainContent}
          </div>
      </NoPrint>
      </PrintProvider>
    );
  }
}

export default App;
