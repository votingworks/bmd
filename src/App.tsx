import Mousetrap from 'mousetrap'
import React from 'react'
// @ts-ignore - @types/react-gamepad doesn't exist
import Gamepad from 'react-gamepad'
import { BrowserRouter, Route, RouteComponentProps } from 'react-router-dom'

import GLOBALS from './config/globals'

import 'normalize.css'
import './App.css'

import {
  handleGamepadButtonDown,
  handleGamepadKeyboardEvent,
} from './lib/gamepad'

import {
  ActivationData,
  CardData,
  Contests,
  Election,
  ElectionDefaults,
  OptionalElection,
  OptionalVote,
  PartialUserSettings,
  TextSizeSetting,
  UserSettings,
  VoterCardData,
  VotesDict,
} from './config/types'

import { getBallotStyle, getContests } from './utils/election'

import Ballot from './components/Ballot'
import Screen from './components/Screen'
import UploadConfig from './components/UploadConfig'
import BallotContext from './contexts/ballotContext'

import ClerkScreen from './pages/ClerkScreen'
import PollWorkerScreen from './pages/PollWorkerScreen'
import PollsClosedScreen from './pages/PollsClosedScreen'
import ActivationScreen from './pages/ActivationScreen'
import CastBallotPage from './pages/CastBallotPage'

import electionDefaults from './data/electionDefaults.json'
import electionSample from './data/electionSample.json'

export const mergeWithDefaults = (
  election: Election,
  defaults: ElectionDefaults = electionDefaults
) => ({ ...defaults, ...election })

interface State {
  ballotStyleId: string
  cardData?: CardData
  contests: Contests
  election: OptionalElection
  isClerkCardPresent: boolean
  isLiveMode: boolean
  isPollsOpen: boolean
  isPollWorkerCardPresent: boolean
  isVoterCardPresent: boolean
  isVoterCardInvalid: boolean
  isRecentVoterPrint: boolean
  machineId: string
  precinctId: string
  ballotsPrintedCount: number
  userSettings: UserSettings
  votes: VotesDict
}

export const electionStorageKey = 'election'
export const activationStorageKey = 'activation'
export const votesStorageKey = 'votes'
export const stateStorageKey = 'state'
const removeElectionShortcuts = ['mod+k']

const initialCardPresentState = {
  isClerkCardPresent: false,
  isPollWorkerCardPresent: false,
  isVoterCardPresent: false,
  isVoterCardInvalid: false,
  isRecentVoterPrint: false,
}

const initialUserState = {
  ballotStyleId: '',
  contests: [],
  precinctId: '',
  userSettings: { textSize: GLOBALS.TEXT_SIZE as TextSizeSetting },
  votes: {},
}

const initialState = {
  ...initialUserState,
  ...initialCardPresentState,
  ballotsPrintedCount: 0,
  election: undefined,
  isLiveMode: false,
  isPollsOpen: false,
  machineId: '---',
}

interface CompleteCardData {
  cardData: CardData
  longValueExists: boolean
}

let checkCardInterval = 0

export class App extends React.Component<RouteComponentProps, State> {
  public state: State = initialState

  public processVoterCardData = (voterCardData: VoterCardData) => {
    const { election } = this.state
    if (!election) {
      return
    }

    const ballotStyle = getBallotStyle({
      ballotStyleId: voterCardData.bs,
      election,
    })
    const precinct = election.precincts.find(pr => pr.id === voterCardData.pr)

    if (ballotStyle && precinct) {
      this.activateBallot({
        ballotStyle,
        precinct,
      })
    } else {
      this.setState({
        isVoterCardInvalid: true,
      })
    }
  }

  public fetchElection = async () => {
    fetch('/card/read_long')
      .then(result => result.json())
      .then(election => {
        this.setElection(JSON.parse(election.longValue))
      })
  }

  public processCardData = (completeCardData: CompleteCardData) => {
    const { cardData, longValueExists } = completeCardData
    switch (cardData.t) {
      case 'voter': {
        const voterCardData = cardData as VoterCardData
        const isBallotPrinted = Boolean(voterCardData.bp)
        const ballotUsedTime = Number(voterCardData.uz) || 0
        const isVoterCardInvalid = Boolean(ballotUsedTime)
        const expirationGracePeriod = 60 * 1000 // 1 minute
        const isRecentVoterPrint =
          isBallotPrinted &&
          ballotUsedTime + expirationGracePeriod > new Date().getTime()
        this.setState({
          ...initialCardPresentState,
          isVoterCardPresent: true,
          isVoterCardInvalid,
          isRecentVoterPrint,
        })
        if (!isVoterCardInvalid) {
          this.processVoterCardData(voterCardData)
        }
        break
      }
      case 'pollworker': {
        // poll worker admin screen goes here
        this.setState({
          ...initialCardPresentState,
          isPollWorkerCardPresent: true,
        })
        break
      }
      case 'clerk': {
        if (longValueExists) {
          this.setState({
            ...initialCardPresentState,
            isClerkCardPresent: true,
          })
        }
        break
      }
    }
  }

  public startPolling = () => {
    if (checkCardInterval === 0) {
      let lastCardDataString = ''

      checkCardInterval = window.setInterval(() => {
        fetch('/card/read')
          .then(result => result.json())
          .then(card => {
            const currentCardDataString = JSON.stringify(card)
            if (currentCardDataString === lastCardDataString) {
              return
            }
            lastCardDataString = currentCardDataString

            if (!card.present || !card.shortValue) {
              this.resetBallot()
              return
            }

            const cardData = JSON.parse(card.shortValue) as CardData
            this.processCardData({
              cardData: cardData,
              longValueExists: card.longValueExists,
            })
          })
          .catch(() => {
            this.resetBallot()
            lastCardDataString = ''
            // if it's an error, aggressively assume there's no backend and stop hammering
            this.stopPolling()
          })
      }, 200)
    }
  }

  public stopPolling = () => {
    window.clearInterval(checkCardInterval)
    checkCardInterval = 0 // To indicate setInterval is not running.
  }

  public markVoterCardUsed = async (
    { ballotPrinted } = { ballotPrinted: true }
  ) => {
    // this is a demo with no card reader attached
    // TODO: limit this to demo elections
    // https://github.com/votingworks/bmd/issues/390
    if (!this.state.isVoterCardPresent) {
      return true
    }
    this.stopPolling()

    const { ballotStyleId, precinctId } = this.getBallotActivation()

    const newCardData: VoterCardData = {
      bs: ballotStyleId,
      pr: precinctId,
      t: 'voter',
      uz: new Date().getTime(),
      bp: ballotPrinted ? 1 : 0,
    }

    const newCardDataSerialized = JSON.stringify(newCardData)

    await fetch('/card/write', {
      method: 'post',
      body: newCardDataSerialized,
      headers: { 'Content-Type': 'application/json' },
    })

    const readCheck = await fetch('/card/read')
    const readCheckObj = await readCheck.json()

    if (readCheckObj.shortValue !== newCardDataSerialized) {
      this.resetBallot()
      return false
    }
    return true
  }

  public componentDidMount = () => {
    if (window.location.hash === '#sample') {
      this.setState({
        election: mergeWithDefaults(electionSample as Election),
      })
    } else {
      const election = this.getElection()
      const { ballotStyleId, precinctId } = this.getBallotActivation()
      const {
        ballotsPrintedCount = initialState.ballotsPrintedCount,
        isLiveMode = initialState.isLiveMode,
        isPollsOpen = initialState.isPollsOpen,
      } = this.getStoredState()
      const ballotStyle =
        ballotStyleId &&
        election &&
        getBallotStyle({
          ballotStyleId,
          election,
        })
      const contests =
        ballotStyle && election
          ? getContests({ ballotStyle, election })
          : initialState.contests
      this.setState({
        ballotsPrintedCount,
        ballotStyleId,
        contests,
        election,
        isLiveMode,
        isPollsOpen,
        precinctId,
        votes: this.getVotes(),
      })
    }
    Mousetrap.bind(removeElectionShortcuts, this.unconfigure)
    document.addEventListener('keydown', handleGamepadKeyboardEvent)
    document.documentElement.setAttribute('data-useragent', navigator.userAgent)
    this.setDocumentFontSize()

    fetch('/machine-id').then(response => {
      response
        .json()
        .catch(() => {})
        .then(responseJSON => {
          if (responseJSON) {
            const { machineId } = responseJSON
            this.setState({
              machineId,
            })
          }
        })
    })

    this.startPolling()
  }

  public componentWillUnount = /* istanbul ignore next - triggering keystrokes issue - https://github.com/votingworks/bmd/issues/62 */ () => {
    Mousetrap.unbind(removeElectionShortcuts)
    document.removeEventListener('keydown', handleGamepadKeyboardEvent)
    this.stopPolling()
  }

  public getElection = (): OptionalElection => {
    const election = window.localStorage.getItem(electionStorageKey)
    return election ? JSON.parse(election) : undefined
  }

  public setElection = (electionConfigFile: Election) => {
    const election = mergeWithDefaults(electionConfigFile)
    this.setState({ election })
    window.localStorage.setItem(electionStorageKey, JSON.stringify(election))
  }

  public getBallotActivation = () => {
    const voterData = window.localStorage.getItem(activationStorageKey)
    return voterData ? JSON.parse(voterData) : {}
  }

  public setBallotActivation = (data: {
    ballotStyleId: string
    precinctId: string
  }) => {
    /* istanbul ignore else */
    if (process.env.NODE_ENV !== 'production') {
      window.localStorage.setItem(activationStorageKey, JSON.stringify(data))
    }
  }

  public getVotes = () => {
    const votesData = window.localStorage.getItem(votesStorageKey)
    return votesData ? JSON.parse(votesData) : {}
  }

  public setVotes = (votes: VotesDict) => {
    /* istanbul ignore else */
    if (process.env.NODE_ENV !== 'production') {
      window.localStorage.setItem(votesStorageKey, JSON.stringify(votes))
    }
  }

  public resetVoterData = () => {
    window.localStorage.removeItem(activationStorageKey)
    window.localStorage.removeItem(votesStorageKey)
  }

  public unconfigure = /* istanbul ignore next - triggering keystrokes issue - https://github.com/votingworks/bmd/issues/62 */ () => {
    this.setState(initialState)
    window.localStorage.clear()
    this.props.history.push('/')
  }

  public setStoredState = () => {
    const { ballotsPrintedCount, isLiveMode, isPollsOpen } = this.state
    window.localStorage.setItem(
      stateStorageKey,
      JSON.stringify({
        ballotsPrintedCount,
        isLiveMode,
        isPollsOpen,
      })
    )
  }

  public getStoredState = () => {
    const storedState = window.localStorage.getItem(stateStorageKey)
    return storedState ? JSON.parse(storedState) : {}
  }

  public updateVote = (contestId: string, vote: OptionalVote) => {
    this.setState(
      prevState => ({
        votes: { ...prevState.votes, [contestId]: vote },
      }),
      () => {
        this.setVotes(this.state.votes)
      }
    )
  }

  public resetBallot = (path: string = '/') => {
    this.resetVoterData()
    this.setState(
      {
        ...initialCardPresentState,
        ...initialUserState,
      },
      () => {
        this.setStoredState()
        this.props.history.push(path)
      }
    )
    this.startPolling()
  }

  public activateBallot = ({ ballotStyle, precinct }: ActivationData) => {
    const { election } = this.state
    if (election) {
      this.setBallotActivation({
        ballotStyleId: ballotStyle.id,
        precinctId: precinct.id,
      })
      this.setState({
        ballotStyleId: ballotStyle.id,
        contests: getContests({ ballotStyle, election }),
        precinctId: precinct.id,
      })
    }
  }

  public setUserSettings = (partial: PartialUserSettings) => {
    this.setState(
      prevState => ({
        userSettings: { ...prevState.userSettings, ...partial },
      }),
      () => {
        const { textSize } = partial
        const isValidTextSize =
          'textSize' in partial &&
          typeof textSize === 'number' &&
          textSize >= 0 &&
          textSize <= GLOBALS.FONT_SIZES.length - 1
        /* istanbul ignore else */
        if (isValidTextSize) {
          this.setDocumentFontSize(textSize!)
        }
      }
    )
  }

  public setDocumentFontSize = (textSize: number = GLOBALS.TEXT_SIZE) => {
    document.documentElement.style.fontSize = `${GLOBALS.FONT_SIZES[textSize]}px`
  }

  public toggleLiveMode = () => {
    this.setState(
      prevState => ({
        isLiveMode: !prevState.isLiveMode,
        ballotsPrintedCount: initialState.ballotsPrintedCount,
        isPollsOpen: initialState.isPollsOpen,
      }),
      this.setStoredState
    )
  }

  public togglePollsOpen = () => {
    this.setState(
      prevState => ({ isPollsOpen: !prevState.isPollsOpen }),
      this.setStoredState
    )
  }

  public incrementBallotsPrintedCount = () => {
    this.setState(
      prevState => ({
        ballotsPrintedCount: prevState.ballotsPrintedCount + 1,
      }),
      this.setStoredState
    )
  }

  public componentDidCatch() {
    this.unconfigure()
    window.location.reload()
  }

  public render() {
    const {
      ballotsPrintedCount,
      ballotStyleId,
      contests,
      election,
      isClerkCardPresent,
      isLiveMode,
      isPollsOpen,
      isPollWorkerCardPresent,
      isVoterCardPresent,
      isVoterCardInvalid,
      isRecentVoterPrint,
      machineId,
      precinctId,
      userSettings,
      votes,
    } = this.state
    if (isClerkCardPresent) {
      return (
        <ClerkScreen
          ballotsPrintedCount={ballotsPrintedCount}
          election={election}
          fetchElection={this.fetchElection}
          isLiveMode={isLiveMode}
          unconfigure={this.unconfigure}
          toggleLiveMode={this.toggleLiveMode}
        />
      )
    } else if (election && isPollWorkerCardPresent) {
      return (
        <PollWorkerScreen
          ballotsPrintedCount={ballotsPrintedCount}
          election={election}
          isLiveMode={isLiveMode}
          isPollsOpen={isPollsOpen}
          machineId={machineId}
          togglePollsOpen={this.togglePollsOpen}
        />
      )
    } else if (election && !isPollsOpen) {
      return <PollsClosedScreen election={election} isLiveMode={isLiveMode} />
    } else if (election) {
      if (isRecentVoterPrint && isVoterCardInvalid) {
        // explicit clicking on this screen for audio because
        // it is being rendered in a different way than the other screens
        // which prevents the typical triggering we've used.
        window.setTimeout(() => {
          const elementToFocus = document.getElementById('audiofocus')
          if (elementToFocus) {
            elementToFocus.focus()
            elementToFocus.click()
          }
        }, 250)
        return <CastBallotPage />
      } else if (
        !isVoterCardInvalid &&
        isVoterCardPresent &&
        ballotStyleId &&
        precinctId
      ) {
        return (
          <Gamepad onButtonDown={handleGamepadButtonDown}>
            <BallotContext.Provider
              value={{
                activateBallot: this.activateBallot,
                ballotStyleId,
                contests,
                election,
                incrementBallotsPrintedCount: this.incrementBallotsPrintedCount,
                isLiveMode,
                markVoterCardUsed: this.markVoterCardUsed,
                precinctId,
                resetBallot: this.resetBallot,
                setUserSettings: this.setUserSettings,
                updateVote: this.updateVote,
                userSettings,
                votes,
              }}
            >
              <Ballot />
            </BallotContext.Provider>
          </Gamepad>
        )
      } else {
        return (
          <ActivationScreen
            election={election}
            isLiveMode={isLiveMode}
            isVoterCardInvalid={isVoterCardInvalid}
          />
        )
      }
    } else {
      return <UploadConfig setElection={this.setElection} />
    }
  }
}

const Root = () => (
  <BrowserRouter>
    <Screen>
      <Route path="/" component={App} />
    </Screen>
  </BrowserRouter>
)

export default Root
