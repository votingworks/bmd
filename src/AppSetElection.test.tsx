import React from 'react'
import { fireEvent, render, wait, waitForElement } from 'react-testing-library'

import electionSample from './data/electionSample.json'

import App from './App'

beforeEach(() => {
  window.localStorage.clear()
  window.location.href = '/'
})

it(`basic end-to-end flow`, async () => {
  /* tslint:disable-next-line */
  const eventListenerCallbacksDictionary: any = {}
  window.addEventListener = jest.fn((event, cb) => {
    eventListenerCallbacksDictionary[event] = cb
  })
  window.print = jest.fn(() => {
    eventListenerCallbacksDictionary.afterprint()
  })

  const { container, getByTestId, getByText } = render(<App />)
  expect(container).toMatchSnapshot()

  /// Upload Config
  const fileInput = getByTestId('file-input')
  fireEvent.change(fileInput, {
    target: {
      files: [
        new File([JSON.stringify(electionSample)], 'election.json', {
          type: 'application/json',
        }),
      ],
    },
  })
  await waitForElement(() => getByText('Scan Your Activation Code'))
  expect(container).toMatchSnapshot()

  // Activation Page
  // TODO: onBlur causes stack overflow error
  // fireEvent.blur(getByTestId('activation-code'))
  fireEvent.change(getByTestId('activation-code'), {
    target: { value: 'Invalid Activation Code' },
  })
  expect(
    (getByTestId('activation-code') as HTMLInputElement).value ===
      'Invalid Activation Code'
  ).toBeTruthy()
  wait(() => (getByTestId('activation-code') as HTMLInputElement).value === '')
  fireEvent.change(getByTestId('activation-code'), {
    target: { value: 'VX.precinct-23.12D' },
  })

  // TODO: replace next line with "Enter" keyDown on activation code input
  fireEvent.click(getByText('Submit'))
  expect(container.firstChild).toMatchSnapshot()
})
