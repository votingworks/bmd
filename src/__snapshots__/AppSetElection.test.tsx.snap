// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sets election config file 1`] = `
.c7 {
  border: none;
  border-radius: 0.25rem;
  box-sizing: border-box;
  background: rgb(211,211,211);
  cursor: pointer;
  padding: 0.75rem 1rem;
  line-height: 1.25;
  color: black;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: auto;
}

.c3 {
  margin: auto auto;
  max-width: 35rem;
  padding: 1rem;
}

.c4 {
  margin: auto;
  width: 100%;
  max-width: 66ch;
  text-align: center;
  line-height: 1.2;
}

.c4 h1 {
  margin: 2rem 0 1rem;
  line-height: 1.1;
  font-size: 1.5rem;
}

.c4 h2 {
  margin: 1.5rem 0 0.75rem;
  font-size: 1.25rem;
}

.c4 h3,
.c4 p {
  margin-top: 1rem;
  margin-bottom: 1rem;
  font-size: 1rem;
}

.c4 h1 + h2 {
  margin-top: -0.75rem;
}

.c4 h1 + p,
.c4 h2 + p {
  margin-top: -0.75rem;
}

.c4 h3 + p {
  margin-top: -1rem;
}

.c4 > :first-child {
  margin-top: 0;
}

.c4 > :last-child {
  margin-bottom: 0;
}

.c6 {
  text-align: center;
  font-size: 0.8rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100%;
}

.c0:focus {
  outline: none;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin: auto;
  width: 100%;
}

.c5 {
  padding: 1rem;
}

@media print {
  .c1 {
    -webkit-box-pack: start;
    -webkit-justify-content: flex-start;
    -ms-flex-pack: start;
    justify-content: flex-start;
    padding: 0;
  }
}

@media print {
  .c3 {
    margin: 0;
    max-width: 100%;
  }
}

@media (min-width:480px) {
  .c4 {
    line-height: 1.3;
  }
}

<div>
  <div
    class="c0"
    tabindex="-1"
  >
    <main
      class="c1"
    >
      <div
        class="c2"
        data-testid="dropzone"
        for="election-file-upload"
        tabindex="0"
      >
        <div
          class="c3"
        >
          <input
            accept="application/json"
            autocomplete="off"
            data-testid="file-input"
            id="election-file-upload"
            style="display: none;"
            tabindex="-1"
            type="file"
          />
          <div
            class="c4"
          >
            <h1>
              Load Election Configuration File
            </h1>
            <p>
              Drag and drop 
              <code>
                election.json
              </code>
               file here, or click to browse for file.
            </p>
            
          </div>
        </div>
      </div>
    </main>
    <footer
      class="c5"
    >
      <div
        class="c4"
      >
        <p
          class="c6"
        >
          <a
            href="/data/election.json"
          >
            Download sample 
            <code>
              election.json
            </code>
             file
          </a>
           
          to upload, or
           
          <button
            class="c7"
            type="button"
          >
            Load Sample Election File
          </button>
        </p>
      </div>
    </footer>
  </div>
</div>
`;
